### PR TITLE
mfa: cancel TOTP prompt if U2F was used

### LINKED ISSUE #5804: mfa: failed registering multiple OTP devices
### Description

**What happened**:

Registering a new OTP device when a user already has one OTP and one U2F device fails:
```
$ tsh mfa add
Choose device type [TOTP, U2F]: totp
Enter device name: otp2
Tap any *registered* security key or enter a code from a *registered* OTP device:
Open your TOTP app and create a new manual entry with these fields:
Name: awly@localhost:3080
Issuer: Teleport
Algorithm: SHA1
Number of digits: 6
Period: 30s
Secret: <redacted>

Once created, enter an OTP code generated by the app: 443161

ERROR:
	rpc error: code = PermissionDenied desc = failed to validate TOTP code: Input length unexpected
```

**What you expected to happen**:

Adding a 2nd OTP device succeeds.

### Details

This happens because the **first** prompt for the OTP code is never canceled (when user taps a U2F token instead). Entering the OTP code for the new device then forwards that input to the previous prompt.
Hitting enter again after that sends an empty input to the 2nd prompt, which fails registration.

https://github.com/gravitational/teleport/blob/master/lib/utils/prompt/confirmation.go#L19-L20 is the relevant TODO.

### PR DESCRIPTION
**Pull request by awly**
Implement context-based cancellation in `/lib/utils/prompt`, for MFA
prompts.

This fixes the following scenario:
```sh
# User has both OTP and U2F devices registered.
$ tsh mfa ls
Name  Type Added at                      Last used
----- ---- ----------------------------- -----------------------------
otp   TOTP Wed, 21 Apr 2021 19:41:44 UTC Wed, 21 Apr 2021 19:44:32 UTC
usb-a U2F  Wed, 21 Apr 2021 19:44:34 UTC Wed, 21 Apr 2021 19:44:34 UTC

# Add a new OTP device, using existing U2F device:
$ tsh mfa add
Choose device type [TOTP, U2F]: totp
Enter device name: otp2
Tap any *registered* security key or enter a code from a *registered* OTP device: <tap> # <- First OTP prompt here
Open your TOTP app and create a new manual entry with these fields:
Name: awly@localhost:3080
Issuer: Teleport
Algorithm: SHA1
Number of digits: 6
Period: 30s
Secret: 3UD42X2NN7EEZ6LUPG6NFMNOLDY6AJTS

Once created, enter an OTP code generated by the app: 607738 # <- Second OTP prompt here
MFA device "otp2" added.
```

Before this PR, the first OTP prompt (for `*registered* device`) would
hang in the background. The OTP code from the newly-registered device is
prompted later, but any text written ends up going to the first prompt and lost.

After this PR, the first prompt is canceled and the code from a new
device goes to the second prompt as intended.

Note: this is implemented using pure Go code (background goroutine
consuming `os.Stdin`) rather than syscalls (e.g. `poll` or `select`)
for portability.

Fixes https://github.com/gravitational/teleport/issues/5804
Updates https://github.com/gravitational/teleport/issues/5878

### CODE REVIEW

File: lib/utils/prompt/stdin.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-r.close:
+		// Close was called, unblock immediately.
+		// r.data might still be blocked if it's blocked on the Read call.
+		return nil, r.getErr()
+	case buf, ok := <-r.data:
+		if !ok {
+			// r.data was closed, so the read goroutine has finished.
+			// No more data will be available, return  the latest error.
```
Comment (Joerger) [inline lib/utils/prompt/stdin.go:125]:
```suggestion
			// No more data will be available, return the latest error.
```

**Review by Joerger — APPROVED** [review]

File: lib/utils/prompt/confirmation.go [hunk 2]
```diff
@@ -33,13 +30,15 @@ import (
...
+//
+// ctx can be canceled to abort the prompt.
+func Confirmation(ctx context.Context, out io.Writer, in *ContextReader, question string) (bool, error) {
 	fmt.Fprintf(out, "%s [y/N]: ", question)
-	scan := bufio.NewScanner(in)
-	if !scan.Scan() {
-		return false, trace.WrapWithMessage(scan.Err(), "failed reading prompt response")
+	answer, err := in.ReadContext(ctx)
+	if err != nil {
+		return false, trace.WrapWithMessage(err, "failed reading prompt response")
```
Comment (r0mant) [inline lib/utils/prompt/confirmation.go:39]: Nit: This code is quite old but I think all "WrapWithMessage" calls here can be replaced with just regular "Wrap".

File: lib/utils/prompt/stdin.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	defer r.mu.Unlock()
+	if r.err != nil {
+		// Keep only the first encountered error.
+		return
+	}
+	r.err = err
+}
+
+func (r *ContextReader) getErr() error {
+	r.mu.Lock()
```
Comment (r0mant) [inline lib/utils/prompt/stdin.go:85]: Nit: Could probably user rw lock instead of a regular one.

**Review by r0mant — APPROVED** [review]

File: lib/utils/prompt/confirmation.go (hunk 2 above)
Comment (awly) [reply lib/utils/prompt/confirmation.go:39]: `Wrap` just calls `WrapWithMessage` if there are any extra arguments.
I don't see any difference between them, other than a bit of overhead

**Pull request merged**